compiler/noirc_evaluator/src/ssa/opt/remove_unreachable_instructions.rs

```diff
@@ -0,0 +1,430 @@
+        }
+        self
+    }
+}
+
+impl Function {
+    fn remove_unreachable_instructions(&mut self) {
+        let mut dom = DominatorTree::with_function(self);
+
+        // The current block we are currently processing
+        let mut current_block_id = None;
+
+        // Whether the current block instructions were determined to be unreachable
+        let mut current_block_instructions_are_unreachable = false;
+
+        // This is the final set of blocks that we concluded have some unreachable instructions.
+        // At the end we'll zero out their terminators.
+        let mut unreachable_blocks = HashSet::default();
+
+        self.simple_reachable_pre_order_blocks_optimization(|context| {
+            let block_id = context.block_id;
+
+            if current_block_id != Some(block_id) {
+                current_block_id = Some(block_id);
+                current_block_instructions_are_unreachable = unreachable_blocks.contains(&block_id);
+            }
+
+            if current_block_instructions_are_unreachable {
+                context.remove_current_instruction();
+                return;
+            }
+
+            let instruction = context.instruction();
+            let is_unreachable = match instruction {
+                Instruction::Constrain(lhs, rhs, _) => {
+                    let Some(lhs_constant) = context.dfg.get_numeric_constant(*lhs) else {
+                        return;
+                    };
+                    let Some(rhs_constant) = context.dfg.get_numeric_constant(*rhs) else {
+                        return;
+                    };
+                    lhs_constant != rhs_constant
+                }
+                Instruction::ConstrainNotEqual(lhs, rhs, _) => {
+                    let Some(lhs_constant) = context.dfg.get_numeric_constant(*lhs) else {
+                        return;
+                    };
+                    let Some(rhs_constant) = context.dfg.get_numeric_constant(*rhs) else {
+                        return;
+                    };
+                    lhs_constant == rhs_constant
+                }
+                _ => false,
+            };
+
+            if is_unreachable {
+                unreachable_blocks.insert(block_id);
+                current_block_instructions_are_unreachable = true;
+
+                add_successors(block_id, context.dfg, &mut dom, &mut unreachable_blocks);
+            }
+        });
+
+        for block_id in unreachable_blocks {
+            let mut terminator = self.dfg[block_id].take_terminator();
+            terminator.map_values_mut(|value_id| {
+                let typ = self.dfg.type_of_value(value_id);
+                zeroed_value(self, block_id, &typ)
+            });
+            self.dfg[block_id].set_terminator(terminator);
+        }
+    }
+}
+
+/// Adds all of a block's successors to the `unreachable_blocks` set if they are dominated by that block.
+fn add_successors(
```

Comment:
It's not obvious from this description that the algorithm will visit all "descendants" and not just the immediate successors. Maybe you could break it up into `blocks_reachable_from` and then another one to add dominated blocks to the unreachable set. Or `blocks_dominated_by` to return the list of all dominated blocks in the graph, with a second step to add them to the unreachable?